switching all memory qualifier code to work with memory qualifier collections, #3841 review

ArielG-NV opened the pull request.

ArielG-NV committed:
Refactor memory qualifier decorators to be a bit-flag set.

replace GloballyCoherent, ReadOnly, WriteOnly, Volatile, and Restrict memory modifiers and decorations with a bit flag set to more efficiently manage memory qualifiers.

added `restrict` modifier to test to ensure the code works when dropping a `restrict` memory qualifier

jkwak-work commented:
Please make your "title" of commit message short so that it can fit into one line.
As an example, the first commit message on this PR has a following title,
```
add memory qualifier collection support fully (and remove individual …
```

The following words are truncated down to the next line.
```
…memory qualifier usage)
```

I don't know how you write your commit message but when I commit, vi warns me with a different color on the characters when I try to make the title line long.
It is very common to have a short title for commit message; not just for Slang project.

jkwak-work commented on source/slang/slang-ir-inst-defs.h:
```diff
@@ -997,6 +992,9 @@ INST(HighLevelDeclDecoration,               highLevelDecl,          1, 0)
     INST(SPIRVBlockDecoration, spvBlock, 0, 0)
 
+        // Stores flag bits of which memory qualifiers an object has
+    INST(MemoryQualifierCollectionDecoration, MemoryQualifierCollectionDecoration, 1, 0)
```
I don't know where the keyword "collection" came from.
But judging by the implementation here, it looks more like "set".
Can you consider renaming "collection" to "set"?

jkwak-work reviewed (approved):
If possible, please squash all of the commits into one before requesting PR.
And rename the title to
```
Refactor memory qualifier decorators to be a bit-flag set.
```

csyonghe committed:
Allow enum values to be used as generic arguments. (#3874)

* Allow enum values to be used as generic arguments.

* Fix constant folding.

csyonghe committed:
Update user guilde for new features. (#3875)

ArielG-NV committed:
Merge branch 'shader-slang:master' into prefer_memory_qualifier_collection_#3841

csyonghe committed:
Fix the erroneous logic of determining whether or not to emit DepthReplacing. (#3885)

* Fix the erroneous logic of determining whether or not to emit DepthReplacing.

Closes #3884.

* Fix.

* More cleanup.

ArielG-NV committed:
Implement 8.14-8.19 of OpenGL-GLSL specification

The following PR implements 8.14-8.19 of the [OpenGL-GLSL specification](https://registry.khronos.org/OpenGL/specs/gl/GLSLangSpec.4.60.pdf).

Fully implements all functions and built-in type's, resolves https://github.com/shader-slang/slang/issues/3692 for GLSL & SPRI-V targets.

_Notes:_
Testing Tools:
* Fragment shaders cannot test computational results. Only OpCodes are checked for proper emitting.

Implementation Notes:
* SubpassInput requires an unknown image format.
* SubpassInput is disjoint from TextureType: __SubpassImpl (.slang) & SubpassInputType (Compiler) to reduce code generation required.
* SubpassInput required an additional input layout modifier, input_attachment_index, this was added as a new parameter binding attribute. Since the following qualifiers can overlap with different resources (`layout(input_attachment_index = 0, binding = 0, set = 0)`) input_attachment_index is checked for overlapping resource bindings separately from other qualifiers with `LayoutResourceKind::InputAttachmentIndex`.
* `GLSLInputAttachmentIndexLayoutModifier` was added to enforce function parameters only accepting `in` decorated variables.
* `in` decorated variables needed to have emitting modified to allow directly emitting the variable into function calls if used as a parameter, normally Slang has a "global variable" shadow as a "global parameter" through a copy. This does not work and is solved using `GlobalVariableShadowingGlobalParameterDecoration` to build a relationship of "global variable" to "global parameter", we then resolve this relationship and replace "global variable" uses later in compile.
* `AtomicCounterMemory` memory-constraint requires `OpCapability AtomicStorage`, `AtomicStorage` is invalid for Vulkan targets. glslang outputs for `barrier`, `memoryBarrier`, and `groupMemoryBarrier` `AtomicCounterMemory` as a memory constraint. This compiles as valid SPIR-V for Vulkan since `OpCapability AtomicStorage` is not declared. This behavior of glslang is undefined as per [3.31.Capability of the SPIR-V specification](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_capability). We will omit `AtomicCounterMemory` from our barrier calls.

ArielG-NV committed:
Refine tests & add SSBO memory qualifer support

add CHECK's to tests to ensure memory qualifiers emit as intended

added tests and changed code to ensure memory qualifiers work on SSBO objects (SPIR-V & GLSL)

ArielG-NV committed:
Merge branch 'prefer_memory_qualifier_collection_#3841' of https://github.com/ArielG-NV/slang into prefer_memory_qualifier_collection_#3841

ArielG-NV committed:
Merge branch 'master'

Removed redundant `coherent` emit in glsl/spir-v to fix merge

ArielG-NV committed:
add memory qualifiers & fixes.

Add to StructuredBuffer & ByteAddressBuffer `ReadOnly`/NonWritable qualifier.

Fixed an error where ReadOnly->NonReadable & WriteOnly->NonWritable
* Adjusted tests accordingly

Added back the removed `globallycoherent` memory qualifier emit'ing code in hlsl-emit (was incorrectly removed).
